lib/logstash/inputs/s3.rb

```diff
@@ -91,189 +129,165 @@ def register
     elsif @credentials.length == 2
       @access_key_id = @credentials[0]
       @secret_access_key = @credentials[1]
-    else
-      raise ArgumentError.new('Credentials must be of the form "/path/to/file" or ["id", "secret"]')
-    end
-
-    if @access_key_id.nil? or @secret_access_key.nil?
-      raise ArgumentError.new('Missing AWS credentials')
-    end
-
-    if @bucket.nil?
-      raise ArgumentError.new('Missing AWS bucket')
-    end
-
-    if @sincedb_path.nil?
-      if ENV['HOME'].nil?
-        raise ArgumentError.new('No HOME or sincedb_path set')
-      end
-      @sincedb_path = File.join(ENV["HOME"], ".sincedb_" + Digest::MD5.hexdigest("#{@bucket}+#{@prefix}"))
-    end
-
-    s3 = AWS::S3.new(
-      :access_key_id => @access_key_id,
-      :secret_access_key => @secret_access_key,
-      :region => @region_endpoint
-    )
-
-    @s3bucket = s3.buckets[@bucket]
-
-    unless @backup_to_bucket.nil?
-      @backup_bucket = s3.buckets[@backup_to_bucket]
-      unless @backup_bucket.exists?
-        s3.buckets.create(@backup_to_bucket)
-      end
     end
 
-    unless @backup_to_dir.nil?
-      Dir.mkdir(@backup_to_dir, 0700) unless File.exists?(@backup_to_dir)
+    if @credentials
+      s3 = AWS::S3.new(
+        :access_key_id => @access_key_id,
+        :secret_access_key => @secret_access_key,
+        :region => @region
+      )
+    else
+      s3 = AWS::S3.new(aws_options_hash)
     end
+  end
 
-  end # def register
+  public
+  def aws_service_endpoint(region)
+    return { :s3_endpoint => region }
+  end
 
   public
   def run(queue)
-    loop do
-      process_new(queue)
-      sleep(@interval)
+    Stud.interval(@interval) do
+      process_files(queue)
     end
-    finished
```

Comment:
The `finished` method was never executed, the only way to get out of the loop is to raise an exception and the exception is catched higher in the pipeline.

We have to revisit how we handle exception in the pipeline.